File: leet_easy_c++/leet_543/leet_543a.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        dobth(root);
        return diam;
    }
    void dobth(TreeNode* iter){
        if(!iter) return;
        diam = max(diam, leg_depth(iter->left) + leg_depth(iter->right));
        dobth(iter->left);
        dobth(iter->right);
    }
    int leg_depth(TreeNode* iter){
        if(!iter) return 0;
        else return 1 + max(leg_depth(iter->left), leg_depth(iter->right));
    }
private:
    int diam; //instantiated to zero;
};

// this solution was my intuitive first solution. It is correct.  After (and while) writing this out I noticed that dobth() was redundant. 

File: leet_easy_java/leet_543/leet_543a.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    //~ dfs(left) + dfs(right) or dobt(left) or dobt(right)
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null) return 0;
        return Math.max(DFS(root.left) + DFS(root.right), Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right)));
    }
    public int DFS(TreeNode iter){
        if(iter == null) return 0;
        return Math.max(DFS(iter.left), DFS(iter.right)) + 1;
    }
}